cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll

```diff
@@ -0,0 +1,590 @@
+   */
+  cached predicate possibleReason(IRGuardCondition guard) {
+    guard = boundFlowCond(_, _, _, _, _)
+    or
+    guard = eqFlowCond(_, _, _, _, _)
+  }
+
+}
+private import RangeAnalysisCache
+import RangeAnalysisPublic
+
+/**
+ * Gets a condition that tests whether `i` equals `bound + delta` at `use`.
+ *
+ * If the condition evaluates to `testIsTrue`:
+ * - `isEq = true`  : `i == bound + delta`
+ * - `isEq = false` : `i != bound + delta`
+ */
+private IRGuardCondition eqFlowCond(Operand op, Operand bound, int delta,
+  boolean isEq, boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.ensuresEq(compared, bound, delta, op.getInstruction().getBlock(), isEq) and
+    result.controls(bound.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  )
+}
+
+/**
+ * Holds if `op1 + delta` is a valid bound for `op2`.
+ * - `upper = true`  : `op2 <= op1 + delta`
+ * - `upper = false` : `op2 >= op1 + delta`
+ */
+private predicate boundFlowStepSsa(
+  NonPhiOperand op2, Operand op1, int delta, boolean upper, Reason reason
+) {
+  op2.getDefinitionInstruction().getAnOperand().(CopySourceOperand) = op1 and
+  (upper = true or upper = false) and
+  reason = TNoReason() and
+  delta = 0
+  or
+  exists(IRGuardCondition guard |
+    guard = boundFlowCond(op2, op1, delta, upper, _) and
+    reason = TCondReason(guard)
+  )
+}
+
+/**
+ * Gets a condition that tests whether `op` is bounded by `bound + delta`.
+ *
+ * If the condition evaluates to `testIsTrue`:
+ * - `upper = true`  : `op <= bound + delta`
+ * - `upper = false` : `op >= bound + delta`
+ */
+private IRGuardCondition boundFlowCond(NonPhiOperand op, NonPhiOperand bound, int delta, boolean upper,
+  boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controls(op.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber(op.getDefinitionInstruction())
+  )
+  // TODO: strengthening through modulus library
+}
+
+/**
+ * Gets a condition that tests whether `op` is bounded by `bound + delta`.
+ * 
+ * - `upper = true`  : `op <= bound + delta`
+ * - `upper = false` : `op >= bound + delta`
+ */
+private IRGuardCondition boundFlowCondPhi(PhiOperand op, NonPhiOperand bound, int delta, boolean upper,
+  boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controlsEdge(op.getPredecessorBlock().getLastInstruction(), op.getInstruction().getBlock(), testIsTrue) and
+    valueNumber(compared.getDefinitionInstruction()) = valueNumber (op.getDefinitionInstruction())
+  )
+  or
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controls(op.getPredecessorBlock(), testIsTrue) and
```

Comment:
Ah, I see now that this is where the "controls" that I mentioned was missing in `controlsEdge` ended up.